src/error.rs

```diff
@@ -3,33 +3,114 @@
-  pub fn message(&self) -> &str {
-    let msg = unsafe { CStr::from_ptr((*self.0).message) };
+/// Wrapper around xmlErrorPtr.
+/// Some fields have been omitted for simplicity/safety
+#[derive(Debug)]
+pub struct StructuredError {
+  /// Human-friendly error message, lossily converted into UTF-8 from the underlying
+  /// C string. May be `None` if an error message is not provided by libxml2.
+  pub message: Option<String>,
+  /// The error's level
+  pub level: XmlErrorLevel,
+  /// The filename, lossily converted into UTF-8 from the underlying C string.
+  /// May be `None` if a filename is not provided by libxml2, such as when validating
+  /// an XML document stored entirely in memory.
+  pub filename: Option<String>,
+  /// The linenumber, or None if not applicable.
+  pub line: Option<c_int>,
+  /// The column where the error is present, or None if not applicable.
+  pub col: Option<c_int>,
+
+  /// The module that the error came from. See libxml's xmlErrorDomain enum.
+  pub domain: c_int,
+  /// The variety of error. See libxml's xmlParserErrors enum.
+  pub code: c_int,
+}
 
-    msg.to_str().unwrap()
+impl StructuredError {
+  /// Copies the error information stored at `error_ptr` into a new `StructuredError`
+  /// 
+  /// # Safety
+  /// This function must be given a pointer to a valid `xmlError` struct. Typically, you
+  /// will acquire such a pointer by implementing one of a number of callbacks
+  /// defined in libXml which are provided an `xmlError` as an argument.
+  /// 
+  /// This function copies data from the memory `error_ptr` but does not deallocate
+  /// the error. Depending on the context in which this function is used, you may
+  /// need to take additional steps to avoid a memory leak.
+  pub unsafe fn from_raw(error_ptr: *mut bindings::xmlError) -> Self {
+    let error = *error_ptr;
+    let message = StructuredError::ptr_to_string(error.message);
+    let level = XmlErrorLevel::from_raw(error.level);
+    let filename = StructuredError::ptr_to_string(error.file);
+
+    let line = if error.line == 0 {
+      None
+    } else {
+      Some(error.line)
+    };
+    let col = if error.int2 == 0 {
+      None
+    } else {
+      Some(error.int2)
+    };
+
+    StructuredError {
+      message,
+      level,
+      filename,
+      line,
+      col,
+      domain: error.domain,
+      code: error.code,
+    }
   }
 
-  /// Return a raw pointer to the underlying xmlError structure
-  pub fn as_ptr(&self) -> *const bindings::_xmlError {
-    self.0 // we loose the *mut since we own it
+  /// Human-readable informative error message.
+  /// 
+  /// This function is a hold-over from the original bindings to libxml's error
+  /// reporting mechanism. Instead of calling this method, you can access the 
+  /// StructuredError `message` field directly.
+  #[deprecated(since="0.3.3", note="Please use the `message` field directly instead.")]
+  pub fn message(&self) -> &String {
+    self.message.as_ref().unwrap()
```

Comment:
To keep the original `&str` type, I think this is a workable variation:
```rust
pub fn message(&self) -> &str {
  self.message.as_deref().unwrap_or("")
}
```

Reply:
Ah excellent! I was about to push a change where the implementation was `self.message.as_ref().unwrap().as_str()`. I didn't think it was even possible to get *both* the `&str` return type *and* a default value working, since the compiler was quite mad about returning references to local values.

I'll push up that change shortly.